File: my-folder/0329-longest-increasing-path-in-a-matrix/solution.java
class Solution {
    int[][] dp;

    public int longestIncreasingPath(int[][] matrix) {
        int m = matrix.length; // ie which row
        int n = matrix[0].length; // ie which col
        dp = new int[m][n];

        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                longestPath(i, j, matrix);
            }
        }

        int max = 0;
        for (int[] row : dp) {
            for (int cell : row) {
                max = Math.max(max, cell);
                System.out.print(cell + " ");
            }
            System.out.println();
        }
        return max;
    }

    public void longestPath(int i, int j, int[][] matrix) {
        int[][] direction = new int[][] {
                { 1, 0 },
                { -1, 0 },
                { 0, 1 },
                { 0, -1 }
        };

        for (int[] dir : direction) {
            int x = i + dir[0];
            int y = j + dir[1];
            if (x < 0 || x >= dp.length || y < 0 || y >= dp[0].length)
                continue;
            if (matrix[x][y] > matrix[i][j]) {
                if (dp[x][y] == 0)
                    longestPath(x, y, matrix);
                dp[i][j] = Math.max(dp[i][j], dp[x][y] + 1);
            }
        }
        if (dp[i][j] == 0)
            dp[i][j] = 1;
    }
}

File: my-folder/0329-longest-increasing-path-in-a-matrix/solution.cpp
class Solution {
public:
    int longestIncreasingPath(vector<vector<int>>& matrix) {
        int ans = 1;
        int n = matrix.size();
        int m = matrix[0].size();
        vector<vector<int>> ll(
            n, vector<int>(m, 1)); // ll[i][j] = length of longest ENDING here

        queue<pair<int, int>> q;
        pair<int, int> dirs[] = {{1, 0}, {-1, 0}, {0, 1}, {0, -1}};

        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                if (ll[i][j] == 1) {
                    q.push(make_pair(i, j));

                    while (!q.empty()) {
                        auto [a, b] = q.front();
                        q.pop();

                        for (auto [dx, dy] : dirs) {
                            int ax = a + dx;
                            int by = b + dy;

                            if (ax < n && ax >= 0 && by >= 0 && by < m &&
                                matrix[ax][by] > matrix[a][b] &&
                                ll[ax][by] < ll[a][b] + 1) {
                                ll[ax][by] = ll[a][b] + 1;
                                ans = max(ans, ll[ax][by]);
                                q.push(make_pair(ax, by));
                            }
                        }
                    }
                }
            }
        }
        return ans;
    }
};
